contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,781 @@
+# Scheduling Policy
+
+_Status: Draft_
+
+_Authors: @arnaudmz, @yastij_
+
+_Reviewers: @bsalamat, @liggitt_
+
+# Objectives
+
+-  Define the concept of scheduling policies
+-  Propose their initial design and scope
+
+## Non-Goals
+
+-  How taints / tolerations work
+-  How NodeSelector works
+-  How node / pod affinity / anti-affinity rules work
+-  How several schedulers can be used within a single cluster
+-  How priority classes work
+
+# Background
+
+During real-life Kubernetes architecturing we encountered contexts where role-isolation (between administration and simple namespace usage in a multi-tenant context) could be improved. So far, no restriction is possible on toleration, priority class usage, nodeSelector, anti-affinity depending on user permissions (RBAC).
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namepace when
+-  using schedulers,
+-  placing pods on specific nodes (master roles for instance),
+-  using specific priority classes,
+-  expressing pod affinity or anti-affinity rules.
+
+Mandatory (and optionally default) values must also be enforced by scheduling policies in case of:
+-  mutli-arch (amd64, arm64) of multi-os (Linux, Windows) clusters
+-  multi-az / region / failure domain clusters
+
+# Overview
+
+Implementing SchedulingPolicy implies:
+-  Creating a new resource named **SchedulingPolicy** (schedpol)
+-  Creating an **AdmissionController** that dehaves on a deny-all-but basis
+-  Allow SchedulingPolicy to be used by pods using RoleBindings or ClusterRoleBindings
```

Comment:
I think we should consider applying the policies to namespaces. That's more aligned with similar K8s policies, such as quota.

Reply:
The point was to be aligned with the PodSecurityPolicy principles:
-  Apply cluster-wide to ease administrator job when creating new namespaces
-  As soon as it is enforced in AdmissionControlers, it applies in a restrictive way to ensure administrators won't let any leaks in permissions

If I understand well, you seem more found of a non-breaking approach

Reply:
@bsalamat - I'm not sure we want to do that, integrating with RBAC would be better in term of experience (e.g. grant cluster-wide usage on a `SchedulingPolicy`).

cc @tallclair @liggitt

Reply:
I think we should support namespaced policy as well. For example, we want `schedulerB` to be used by only `namespaceB` - How do we restrict it using current policy?

The way I see it, we should have 2 policies or atleast have a field in the spec which tells namespace defaults.

Reply:
@ravisantoshgudimetla - this is all enabled by RBAC (roleBindings allow the verb «use » on a schedulingPolicy on a specific namespace, ClusterRoleBindings on the other hand will allow a cluster-wide usage of a schedulingPolicy, having a namespaceSelector is not viable for this usecase)

Reply:
@ravisantoshgudimetla: to be very precise, enforcing `schedulerB` to be used by only `namespaceB` can be achieved by:

1) create a SchedulingPolicy say `policyB`:

```yaml
apiVersion: extensions/valpha1
kind: SchedulingPolicy
metadata:
  name: policyB
spec:
  allowed:
    schedulerNames: ["default-scheduler", "schedulerB"]
```
2) create a ClusterRole allowing to use policyB:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: policyB
rules:
- apiGroups: ['extensions']
  resources: ['schedulingpolicies']
  verbs:     ['use']
  resourceNames:
  - policyB
```

3) create a RoleBinding that applies to all serviceaccounts in namespace `namespaceB`:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: policyB
  namespace: namespaceB
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: policyB
subjects:
- kind: Group
  name: system:serviceaccounts:namespaceB
  apiGroup: rbac.authorization.k8s.io
```

Other service accounts (in any other namespaces) will fallback to the default `restricted` policy that forbids the use of this scheduler.

Reply:
@arnaudmz @yastij Thanks. Clearly my example is not complex enough. My question is more on the lines of how do we ensure that certain attributes of scheduling like nodeSelector could come from a namespace rather than created by whom and others created by whom. For example as of now there is namespace level whitelist for tolerations. How can we tell that this toleration is valid or not until the namespace exists.

Reply:
The use of RBAC to determine which PodSecurityPolicy applies is one of the most confusing things we've done in the entire system, which is saying a lot.

Another model to consider is NetworkPolicy:
https://kubernetes.io/docs/concepts/services-networking/network-policies/